### PR TITLE
[WIP][Testing] Mock Persistence Module in Utility Tests  - An Example

### LINKED ISSUE #261: [Testing] Mock 10% Persistence Module in Utility Tests
## Objective

Mock the Persistence Module for all Utility Testing

## Origin Document

The Utility Module relies on the Persistence Module in order to execute all of it's testing:

See 
```golang
func newTestPersistenceModule(_ *testing.M, databaseUrl string) modules.PersistenceModule {
```

Under `utility/test`

## Goals

- Enabling utility tests for persistence business logic that doesn't exist
- Remove cross module imports in Utility
- Allow complete encapsulation of test suite in Utility
- Create a clear and understandable `Mock`

## Deliverable

- Code complete implementation
- Plenty of comments in the `Mock`
- Updated changelog

## Non-goals / Non-deliverables

- Little nits that have nothing to do with the issue

## General issue deliverables

- [ ] Update the appropriate CHANGELOG
- [ ] Update any relevant READMEs (local and/or global)
- [ ] Update any relevant global documentation & references
- [ ] If applicable, update the source code tree explanation
- [ ] If applicable, add or update a state, sequence or flowchart diagram using [mermaid](https://mermaid-js.github.io/mermaid/)

## [Optional] Testing Methodology

- \__REPLACE_ME_: Make sure to update the testing methodology appropriately\_
- **Task specific tests**: `make ...`
- **All tests**: `make test_all`
- **LocalNet**: verify a `LocalNet` is still functioning correctly by following the instructions at [docs/development/README.md](https://github.com/pokt-network/pocket/tree/main/docs/development)

---

**Creator**: @andrewnguyen22

### PR DESCRIPTION
**Pull request by Olshansk**
## Description

This PR is a WIP and is meant to be a starting point with an example.

The goal is to enable the utility module to be developed while dependencies in the persistence module are not yet ready. This paradigm could extend to any other cross-module interaction.

The comments contain more detail, but in short:
1. A real instance of the persistence context is instantiated
2. A mocked instance of the persistence module instantiated and mocked
3. For some functions, we implement a passthrough from the mock -> real implementation
4. For other functions, we mock the behaviour; sometimes it's stateless and sometimes it's stateful

## Issue

Fixes #261

## Type of change

Please mark the relevant option(s):

- [x] New feature, functionality or library
- [ ] Bug fix
- [x] Code health or cleanup
- [ ] Major breaking change
- [ ] Documentation
- [ ] Other

## TODO: List of changes

- Change #1
- Change #2
- ...

## Testing
- [x] `go test -v -count=1 -v ./utility/... -run TestExperimental`
- [ ] `make develop_test`
- [ ] [LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md) w/ all of the steps outlined in the `README`

## Required Checklist

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have tested my changes using the available tooling
- [ ] I have updated the corresponding CHANGELOG

### If Applicable Checklist
- [ ] I have updated the corresponding README(s); local and/or global
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added, or updated, [mermaid.js](https://mermaid-js.github.io) diagrams in the corresponding README(s)
- [ ] I have added, or updated, documentation and [mermaid.js](https://mermaid-js.github.io) diagrams in `shared/docs/*` if I updated `shared/*`README(s)

### CODE REVIEW

File: utility/test/module_test.go [hunk 1]
```diff
@@ -67,7 +66,19 @@ func NewTestingUtilityContext(t *testing.T, height int64) utility.UtilityContext
...
+		Mempool:      NewTestingMempool(t),
+		Context: &utility.Context{
+			PersistenceRWContext: persistenceContext,
+			SavePointsM:          make(map[string]struct{}),
+			SavePoints:           make([][]byte, 0),
+		},
+	}
+}
+
+func newTestPersistenceModule(m *testing.M, databaseUrl string) modules.PersistenceModule {
```
Comment (andrewnguyen22) [inline utility/test/module_test.go:81]: Why do we still have the existing persistence module here? Is it safe to assume that this will be removed by this issue?

File: utility/test/experimental_test.go [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+	err = ctx.AddAccountAmount(addr, big.NewInt(42))
+	require.NoError(t, err)
+
+	// Check the app hash
+	appHash, err = ctx.GetAppHash()
+	require.NoError(t, err)
+	require.Equal(t, []byte(" 1  getHash  42  getHash "), appHash)
+}
+
+func newMockablePersistenceContextForAppHashTestWithMoreState(t *testing.T, height int64) modules.PersistenceRWContext {
```
[the remark this answers is not in the document]
Comment (andrewnguyen22) [reply utility/test/experimental_test.go:179]: Conversation around 'full mocking' vs 'hybrid' here https://github.com/pokt-network/pocket/pull/290#issuecomment-1270361284

File: utility/test/experimental_test.go [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+	"testing"
+
+	"github.com/golang/mock/gomock"
+	"github.com/pokt-network/pocket/shared/crypto"
+	"github.com/pokt-network/pocket/shared/modules"
+	modulesMock "github.com/pokt-network/pocket/shared/modules/mocks"
+	"github.com/stretchr/testify/require"
+)
+
+// This test just tries to add amounts to an account and check the app hash via the real implementation
```
[the remark this answers is not in the document]
Comment (Olshansk) [reply utility/test/experimental_test.go:15]: Yea, the way I've seen this done before is that you don't build it out all at once, but you add the functions/components you need over time.

File: utility/test/experimental_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (Olshansk) [reply utility/test/experimental_test.go:179]: > like a 1:1 mapping of sorts?

Doesn't have to be a 1:1 mapping.

You can create helpers and then reuse them:

Imagine a `common_mocks_test.go` file
```go
    func sharedAddAccountAmount(address []byte, amount string) error {
       ... // implementation
    }
    
    // define other common functions
    
    sharedPersistenceContextMock := initializedPersistenceMock()
  
	sharedPersistenceContextMock.EXPECT().
		AddAccountAmount(gomock.Any(), gomock.Any()).
		DoAndReturn(sharedAddAccountAmount).
		AnyTimes()

    // expect other common functions
```

`some_test.go`

```go
    persistenceContextMock = copy(sharedPersistenceContextMock)
    
    // override some subset of mocks specific to this
```

`some_other_test.go`

```go
    persistenceContextMock = copy(sharedPersistenceContextMock)
    
    // override some other subset of mocks specific to this
```

**Comment by Olshansk** [comment]
> This appears to not satisfy the encapsulation requirement

Yea, I see it as a step in the right direction.

> why don't we just Mock everything instead of using the persistence module?

I guess I have varying viewpoints here but don't feel strongly.

1. The persistence module has dozens of functions, and I was thinking that we could leverage existing business logic in the persistence module as an interim to unblock other work dependent on business logic that's not been implemented yet.

2. If we mock with placeholders, it makes sense, but if we mock with "real business logic", then we will pretty much need to maintain two complete versions of the persistence module. E.g., anytime we change something in the real persistence module, we'd need to change that here too. Could be the right path, but feels complex to me straight of the bat.

> It appears (atleast for the unit tests) we would be able to use the Expect statements to be able to test unlimited scenarios

Correct.

---

tl;dr I think what the "passthrough" approach does is unblock work that's dependant on functionality we don't have yet, without needing to pay the upfront cost of doing a complete second implementation of the persistence module. This may or may not be a good idea, but I think we will have visibility into it as we start mocking parts of it and get a clearer picture.

**Comment by andrewnguyen22** [comment]
@Olshansk seems like a decent amount of techdebt will be created if we go passthrough approach. I caution about the snowplow effect that @jessicadaugherty alluded to. 

The current open M3 issues for this sprint are not blocked by this, but the future ones will be. Since this is an XL task, I'd encourage us to go all the way with the Mock

**Comment by Olshansk** [comment]
> Since this is an XL task, I'd encourage us to go all the way with the Mock

To make sure I understand, are you suggesting to maintain two "feature complete" versions of the persistence module, or just mocking everything in the persistence module interface with hardcoded values.

Will ping to sync offline as well.

**Comment by andrewnguyen22** [comment]
> just mocking everything in the persistence module interface with hardcoded values.

I'm suggesting to mock the entire persistence module in the Utility module to fully encapsulate rather than importing the persistence module and passing through

**Comment by deblasis** [comment]
I have been thinking about this and I have a blurred idea that I wanted to share with you guys.

How about a function that constructs a mock and then multiple methods that in a fluent way allow us to configure expected behaviours?

Pseudo-code of a random scenario:

```go
persistenceContextMock:=newPersistenceContextMock(ctrl).
FromRuntimeMgr(runtimeMgr).
ShouldFailWhenGetAllFishermenIsCalledAtHeight(2, fmt.Errorf("some expected failure here").
AppHashShouldReturnAtHeight(2, []byte("somehash"))

persistenceMock:=newPersistenceModuleMock(ctrl).
WithPersistenceContextMock(persistenceContextMock).
SomeOtherBehaviouralMethod(...some_arguments)
```

And these can also be wrapped in some sort of util if we call them all the time and change only a bunch of parameters.

so in `Utility` you would call

```go
persistenceModuleMock := getHappyPathXXPersistenceModule()
```

It might be more complex than it sounds especially because probably there's gonna be some state to be handled inside the mock itself 🤔 but we could leverage the built-in controller to keep track of calls instead. I need to read more in depth how it all works.

**Comment by Olshansk** [comment]
_Heads up: I added the typical Olshansky whitespaces in `utility/test/account_test.go` to help me understand the tests better._

@deblasis I believe the implementation I pushed is a starting point to what you described so PTAL.

@andrewnguyen22 I removed the code for my initial passthrough attempt and updated the unit tests in `utility/test/account_test.go` to use the new approach. 

Note how in places where we have dependencies on accounts, I do:

```go
	runTimeMgr := persistenceRuntimeMgr(t)
	acc := GetAllTestingAccounts(t, ctx)[0]
	ctx := NewTestUtilityContext(t, 0, withBaseAccountMock(t, runTimeMgr))
```

And in some places where we have dependencies on pools, I do:

```go
	runTimeMgr := persistenceRuntimeMgr(t)
	pool := GetAllTestingPools(t, ctx)[0]
	ctx := NewTestUtilityContext(t, 0, withBasePoolMock(t, runTimeMgr))
```

What the current implementation achieves:
1. Complete separation from the persistence module
2. Composable "building blocks" for a stateful persistence module mock that can be used in utility. E.g. you could include both pools AND accounts by running `NewTestUtilityContext(t, 0, withBasePoolMock(t, runTimeMgr), withBaseAccountMock(t, runTimeMgr))` if you needed to.
3. `withBasePoolMock` and `withBaseAccountMock` are stateful mocks, BUT we can explicitly EXPECT other functions with specific (i.e. not `gomock.Any()` values when we have more complex tests verifying specific flows

Please note that I do not believe this is "simple" or "low cognitive load" but fits the requirements of isolation by mocking the `shared` modules everywhere and sets up up to have other building blocks we can add in the future.

Let me know if jumping on a call would help instead.

**Comment by Olshansk** [comment]
@andrewnguyen22 Wanted to bump this for you to provide some feedback if this is what you had in mind.

**Review by andrewnguyen22 — CHANGES_REQUESTED** [review]
DUP COMMENT SEE BELOW

**Comment by Olshansk** [comment]
Closing out this PR for now, but we can use it as a reference for future work on #261.
